repo/content/content_manager_indexes.go

```diff
@@ -93,13 +94,19 @@ func (bm *Manager) compactIndexBlobs(ctx context.Context, indexBlobs []IndexBlob
 	var inputs, outputs []blob.Metadata
 
 	for _, indexBlob := range indexBlobs {
-		if err := bm.addIndexBlobsToBuilder(ctx, bld, indexBlob, opt); err != nil {
+		formatLog(ctx).Debugf("adding index entries %v to compacted index", indexBlob)
+
+		if err := bm.addIndexBlobsToBuilder(ctx, bld, indexBlob); err != nil {
 			return errors.Wrap(err, "error adding index to builder")
 		}
 
 		inputs = append(inputs, indexBlob.Metadata)
 	}
 
+	// after we built index map in memory, drop contents from it
+	// we must do it after all input blobs have been merged, otherwise we may resurrect contents.
+	dropContentsFromBuilder(ctx, bld, opt)
```

Comment:
This solves the case where there are multiple entries (in separate indices) for a given content (`ci`) and one of the entry with the most recent time stamp is marking the content as deleted and the compaction includes **all** the indices that have entries for `ci`.

My assumption was that during a full maintenance **all the indices** would be included, and thus it was safe to drop entries without resurrecting them.

Looking at L75 above (https://github.com/kopia/kopia/pull/563/files#diff-c8900248980c856e920adf28aa8b1019R75) it seems that even when `CompactOptions.AllIndexes == true`, there could be (maybe rare) cases where not all the indices are considered, and thus content could be resurrected from an old (relatively large) index that was not considered during the compaction. Or perhaps I am missing something.

Reply:
Follow up to the comment above.

Let me take that back.
It seems that during full maintenance, `getBlobsToCompact` (https://github.com/kopia/kopia/pull/563/files#diff-c8900248980c856e920adf28aa8b1019R75) would return all blobs because `CompactOptions.MaxSmallBlobs == 0` during full maintenance.

There are a number of "long-range dependencies and assumptions" in how the different code paths interact that make it harder to reason about the invariants maintained by maintenance.